test/components/node-problem-detector/nodeproblemdetector_test.go

```diff
@@ -0,0 +1,163 @@
+}
+
+//nolint: funlen
+func TestNodeProblemDetectorNodeEvent(t *testing.T) {
+	t.Parallel()
+
+	client := testutil.CreateKubeClient(t).CoreV1()
+	nsclient := client.Namespaces()
+
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: testutil.TestNamespace("node-problem-detector"),
+		},
+	}
+
+	privileged := true
+	// Pod config.
+	pod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "node-problem-detector-test-",
+		},
+		Spec: corev1.PodSpec{
+			RestartPolicy: "Never",
+			Containers: []corev1.Container{
+				{
+					Name:  "nginx",
+					Image: "nginx",
+					SecurityContext: &corev1.SecurityContext{
+						Privileged: &privileged,
+					},
+					Command: []string{"sh"},
+					Args:    []string{"-c", "echo 'kernel: BUG: unable to handle kernel NULL pointer dereference at TESTING_IS_COOL' >> /dev/kmsg"}, //nolint:lll
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "kmsg",
+							MountPath: "/dev/kmsg",
+						},
+					},
+				},
+			},
+			Volumes: []corev1.Volume{
+				{
+					Name: "kmsg",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/dev/kmsg",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// Creation of namespace.
+	ns, err := nsclient.Create(context.TODO(), ns, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("creating namespace: %v", err)
+	}
+
+	t.Cleanup(func() {
+		if err := nsclient.Delete(context.TODO(), ns.Name, metav1.DeleteOptions{}); err != nil {
+			t.Logf("failed removing namespace: %v", err)
+		}
+	})
+
+	podsClient := client.Pods(ns.Name)
+
+	// Retry pod creation. This might fail if node-problem-detector is not ready yet and some requests might fail.
+	if err := wait.PollImmediate(testutil.RetryInterval, testutil.TimeoutSlow, func() (done bool, err error) {
+		pod, err = podsClient.Create(context.TODO(), pod, metav1.CreateOptions{})
+		if err != nil {
+			t.Logf("retrying pod creation, failed with: %v", err)
+
+			return false, nil
+		}
+
+		return true, nil
+	}); err != nil {
+		t.Fatalf("creating pod: %v", err)
+	}
+
+	phase := corev1.PodUnknown
+
+	if err := wait.PollImmediate(testutil.RetryInterval, testutil.TimeoutSlow, func() (done bool, err error) {
+		p, err := podsClient.Get(context.TODO(), pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("couldn't get the pod %q: %w", pod.Name, err)
+		}
+
+		// Check for event.
+		events, err := client.Events("default").List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return false, fmt.Errorf("couldn't list the events %w", err)
+		}
+
+		phase = p.Status.Phase
+		if phase == corev1.PodSucceeded || phase == corev1.PodFailed {
+			for _, v := range events.Items {
+				if v.Reason == "KernelOops" && strings.Contains(v.Message, "TESTING_IS_COOL") {
```

Comment:
So node detector puts not only node condition, but also pod events?

Reply:
temporary problems get reported as Events, and permanent problems get reported as Node Conditions.

Reply:
Cool, I didn't know, thanks!